### ISSUE TITLE
slow training of LSTM

### ISSUE THREAD

**Issue by ffmpbgrnn** [fchollet/keras#415]
Hi, I just started using keras. Awesome work!
I tried to use LSTM with the following code
```python
model = Sequential()
model.add(LSTM(4096, 512, return_sequences=True))                                                                                                                
model.add(TimeDistributedDense(512, 4096))
model.add(Activation('time_distributed_softmax'))
model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
```
And I compared the efficiency with [char-rnn](https://github.com/karpathy/char-rnn), and found that the implementation in keras is about 4 times slower than Karpathy's (with the same batchsize).
Am I doing something wrong? 
I've attached the theano profile [result](https://gist.github.com/ffmpbgrnn/842e1910f216b1e00e27)
Thanks you!

**Comment by fchollet** [comment]
Training time is heavily dependent on network size and batch size. Is it even the same network (size included) at all?

Also, time_distributed_softmax is deprecated now, use softmax instead.

**Comment by ffmpbgrnn** [comment]
Yes. I tested with one layer LSTM with the same sequence length, memory cell size, input dimension, output dimension and batch size as well.
Thank you for pointing out the depreacated softmax usage.

**Comment by ffmpbgrnn** [comment]
From the [profile](https://gist.github.com/ffmpbgrnn/842e1910f216b1e00e27), it seems it spent much time on operation: `Elemwise{Composite{(i0 * log((i1 / i2)))}}`

**Comment by fchollet** [comment]
And what batch size are you using?

**Comment by ffmpbgrnn** [comment]
I'm using the batch size of 128.

**Comment by fchollet** [comment]
> From the profile, it seems it spent much time on operation: `Elemwise{Composite{(i0 * log((i1 / i2)))}}`

Thanks for the info. Try setting the `inner_activation` parameter to `theano.tensor.nnet.ultra_fast_sigmoid`, and tell me what you get.

**Comment by ffmpbgrnn** [comment]
Hi, in `activations.py` I added
```python
def ultra_fast_sigmoid(x):                                                                                                                                                
    return T.nnet.ultra_fast_sigmoid(x)
```
and rebuilt keras and then call LSTM with `LSTM(4096, 512, return_sequences=True, inner_activation='ultra_fast_sigmoid')`

However, it failed to compile with the error:
```
Using gpu device 1: Tesla K20m
Build model...
Traceback (most recent call last):
  File "fc_rnn.py", line 40, in <module>
    model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
  File "build/bdist.linux-x86_64/egg/keras/models.py", line 316, in compile
  File "build/bdist.linux-x86_64/egg/keras/optimizers.py", line 75, in get_updates
  File "build/bdist.linux-x86_64/egg/keras/optimizers.py", line 24, in get_gradients
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 547, in grad
    grad_dict, wrt, cost_name)
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1288, in _populate_grad_dict
    rval = [access_grad_cache(elem) for elem in wrt]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1093, in access_term_cache
    input_grads = node.op.grad(inputs, new_output_grads)
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/scan_module/scan_op.py", line 1851, in grad
    _dC_dinps_t = compute_gradient(Xt, dC_dXt)
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/scan_module/scan_op.py", line 1783, in compute_gradient
    return_disconnected='None')
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 547, in grad
    grad_dict, wrt, cost_name)
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1288, in _populate_grad_dict
    rval = [access_grad_cache(elem) for elem in wrt]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 953, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1246, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 1093, in access_term_cache
    input_grads = node.op.grad(inputs, new_output_grads)
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/tensor/elemwise.py", line 672, in grad
    rval = self._bgrad(inputs, ograds)
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/tensor/elemwise.py", line 747, in _bgrad
    scalar_igrads = self.scalar_op.grad(scalar_inputs, scalar_ograds)
  File "/home/archy/.local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/scalar/basic.py", line 890, in grad
    self.__class__.__name__)
theano.gof.utils.MethodNotDefined: ('grad', <class 'theano.tensor.nnet.sigm.UltraFastScalarSigmoid'>, 'UltraFastScalarSigmoid')
```

I'm new to theano and have no idea how to fix this. And I'm running with `THEANO_FLAGS=mode=FAST_RUN,device=gpu1,floatX=float32,nvcc.fastmath=True,allow_gc=False python rnn.py `.
Maybe there's something wrong in [this function](https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/sigm.py#L91)?

**Comment by ffmpbgrnn** [comment]
Hi, I replaced both `sigmoid` and `tanh` activations with `linear`, and get the [profile](https://gist.github.com/ffmpbgrnn/842e1910f216b1e00e27#file-linear-log).

**Comment by fchollet** [comment]
I tried running the same network, but I'm getting completely different profiles (on CPU): 
```
Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  44.8%    44.8%      18.078s       9.42e-02s     C      192       12   Dot22
  14.3%    59.1%       5.754s       3.60e-01s     Py      16        1   forall_inplace,cpu,grad_of_scan_fn}
   6.2%    65.3%       2.503s       1.56e-01s     C       16        1   Elemwise{Composite{(i0 * log((i1 / i2)))}}
   4.9%    70.1%       1.960s       1.02e-02s     C      192       12   Elemwise{Composite{(i0 - ((i1 * i2) / sqrt((i3 + i4 + i5))))}}
   4.8%    74.9%       1.937s       1.21e-01s     Py      16        1   forall_inplace,cpu,scan_fn}
   3.7%    78.7%       1.503s       1.88e-02s     C       80        5   Sum{axis=[0, 1], acc_dtype=float64}
```

**Comment by ffmpbgrnn** [comment]
I profiled on CPU as well, and the [CPU profile](https://gist.github.com/ffmpbgrnn/842e1910f216b1e00e27#file-cpu-log-L42) also shows that the  `Dot22` costs most

**Comment by kylemcdonald** [comment]
i noticed the result here is in a different, larger regime than i've been operating in. so i wanted to share a result that is more in the range that most people have been running char-rnn.

on my machine (nvidia 750m) for the first epoch (228 minibatches) i see:

- char-rnn is 0.20s/minibatch
- keras is 0.28s/minibatch

this is using the following parameters (based on char-rnn's defaults):

- using the nietzsche dataset
- 128 units per layers
- 2 layers
- 2e-3 learning rate
- 0.95 decay rate for rmsprop
- 50 token sequence length
- 50 token step size (no overlap)
- 50 "sentence" batch size
- gradient clip at 5
- no dropout (dropout did not seem to have a significant effect for char-rnn or keras) 

i haven't done any profiling. this is just based on modifying `lstm_text_generation.py` to more closely match char-rnn https://github.com/kylemcdonald/keras/blob/char_rnn/examples/char_rnn.py

i also haven't compared the loss curves between keras and char-rnn after a fixed number of iterations. i'll try that next. presumably they will vary a bit based on other internal parameters...

**Comment by ffmpbgrnn** [comment]
@kylemcdonald , in the original version of [char-rnn](https://github.com/karpathy/char-rnn/), it predicts character by character, not predicting the next character of a sequence (in your case 50). Your version of `# the last layer has return_sequences=False` might be not comparable to the original char-rnn?

**Comment by fchollet** [comment]
> 50 "sentence" batch size

Playing around with the batch size might allow you to get some speedup on GPU. Batch sizes that are too small mean that a lot of time will be spent moving minibatch data to and from the GPU memory.

> Your version of # the last layer has return_sequences=False might be not comparable to the original char-rnn?

So how does char-rnn work exactly? I must say I've never looked at it. In any case I would not expect much computational difference between `return_sequences=False` and `return_sequences=True` since in both cases the full sequence is being computed anyway. The computation steps are the exact same.

**Comment by kylemcdonald** [comment]
@ffmpbgrnn regarding `return_sequences`, this could also be done with `True` for all layers and then using a non-zero index on the `predict()` result to get the last index. i don't think it's really changing anything, as @fchollet explains above.

but i think i understand what you're saying about the prediction being different. char-rnn samples one character at a time without resetting the state of the network. the keras lstm example, however, feeds in a `maxlen` sequence for every step, presumably resetting the state of the net before each prediction? meaning any relationships longer than `maxlen` will absolutely not be captured. (if `fit()` is also resetting the state of the network, this is a moot point.)

i tried modifying the prediction process to see if anything happens:

```python
        x = np.zeros((1, len(sentence), len(chars)))
        for t, char in enumerate(sentence):
            x[0, t, char_indices[char]] = 1.
        preds = model.predict(x, verbose=0)[0]
        next_index = sample(preds, diversity)
        next_char = indices_char[next_index]
        sys.stdout.write(next_char)

        for iteration in range(400):
            x = np.zeros((1, 1, len(chars)))
            x[0, 0, next_index] = 1.

            preds = model.predict(x, verbose=0)[0]
            next_index = sample(preds, diversity)
            next_char = indices_char[next_index]
...
```

but this just spits out complete gibberish. this makes me think the state is reset for every `predict()`.

here is the comparison of the loss curves:

<img width="604" alt="screen shot 2015-07-25 at 00 20 25" src="https://cloud.githubusercontent.com/assets/157106/8887863/2b852bb6-3263-11e5-94ef-78f8436de829.png">

keras drops very low, and char-rnn bottoms out somewhere around epoch 20 (i don't have accuracy numbers for char-rnn).

but this graph doesn't seem to capture what's actually being learned. there's definitely something different that char-rnn is doing compared to the keras lstm example. after 20 epochs, char-rnn prints:

```
$ th sample.lua cv/lstm_19.00e_2x128_0.00d_50_1.4290l.t7 -temperature 1
using CUDA on GPU 0...	
creating an LSTM...	
missing seed text, using uniform probability over first character	
--------------------------	
god are questions has not already capable of the inner favourite confess or wither finders and delights--and HOPEOTOL, and then seeks he is always was even for his keens to have growt in all the science, all, in the soul of--home in so Asuated equalize, more ppodenus, that
existence, and are natured to forth and something of the lake
religious spoten of heaused by merit
of the desire them--our ofteic plumsing expluaring heredatical, relicy also, and more
deceivative
ciriums, "I purpfective excitable, it is note-side any greatred of reason
when consequently
```

etc.

but keras prints:

```
----- diversity: 1.0
----- Generating with seed: "o rich in futurity. This kind of music
expresses b"
o rich in futurity. This kind of music
expresses be intlest it of the to surm of the mort eans and no
btill se paa antate
fark, ind woord peans. 
W
 

PoWPo)t -eeds aptel cist of beevo aloxhe coatica abol?ess, and to tame of thes to purf so:t arr-istmentated therrerepe then wath imsold to tweresce is is and in wa kesove to as mort that elustion. Nut whac allest a sukould and Bglol,. -whe worde dears
or the world, and jorbend selt atse ald alags
o
```

to me, the keras output looks a lot like the char-rnn output before the network is "warmed up" by outputting enough data. compare how many words are "real" between the two samples, and the distance between spaces.

in summary: if there is a way to run `fit()` and `predict()` without resetting the state of the network (which is presumably happening behind the scenes, but i can't find where) then the results might be more similar to char-rnn. and regarding this issue, it could explain the relative speed of the sampling compared to char-rnn, but not the speed of `fit()`.

finally: thanks @fchollet for the great work on this library, for me it's at the perfect level of abstraction for the level i'm at right now, and i love the way it's designed with callbacks and lambda functions where necessary :)

**Comment by fchollet** [comment]
@kylemcdonald it's an open issue: https://github.com/fchollet/keras/issues/98#issuecomment-122776799

I expect we'll support it within the next couple weeks. I don't think it would be hard to do, we just need somebody to look into it and, you know, you just do it. I've been busy with other things so far.

**Comment by kylemcdonald** [comment]
aha, great! i read through #98 hoping to be able to help, but it seems that implementing stateful training & prediction requires a deeper understanding of theano than i have right now. so i'll step back and watch from the sidelines :)
